langchain/alignment/guards.py

```diff
@@ -0,0 +1,314 @@
+"""A utilities for guarding the output of a given chain or agent.
+"""
+
+from difflib import SequenceMatcher
+from typing import Any, Callable, List, Tuple
+
+from langchain.chains import LLMChain
+from langchain.prompts import PromptTemplate
+from langchain.utilities.normalization import normalize_boolean_output
+
+
+class BaseGuard:
+    """The Guard class is a decorator that can be applied to any chain or agent to either throw an error or recursively call the chain or agent when the output of said chain or agent violates the rules of the guard. The BaseGuard alone does nothing but can be subclassed and the resolve_guard function overwritten to create more specific guards.
+
+    Args:
+        retries (int, optional): The number of times the chain or agent should be called recursively if the output violates the restrictions. Defaults to 0.
+
+    Raises:
+        Exception: If the output violates the restrictions and the maximum number of retries has been exceeded.
+
+    """
+
+    def __init__(self, retries: int = 0, *args: Any, **kwargs: Any) -> None:
+        self.retries = retries
+
+    def resolve_guard(
+        self, llm_response: str, *args: Any, **kwargs: Any
+    ) -> Tuple[bool, str]:
+        """Function to determine if guard was violated (if response should be blocked)
+        Can be overwritten when subclassing Guard class to expand on guard functionality
+
+        Args:
+            llm_response (str): the llm_response string to be tested against the guard.
+
+        Returns:
+            tuple:
+                bool: True if guard was violated, False otherwise.
+                str: The message to be displayed when the guard is violated (if guard was violated).
+        """
+        return False, ""
+
+    def handle_violation(self, message: str, *args: Any, **kwargs: Any) -> Exception:
+        """Function to handle violation of guard.
+
+        Args:
+            message (str): the message to be displayed when the guard is violated.
+
+        Raises:
+            Exception: the message passed to the function.
+        """
+        raise Exception(message)
+
+    def __call__(self, func: Callable) -> Callable:
+        def wrapper(*args: Any, **kwargs: Any) -> str:
+            if self.retries < 0:
+                raise Exception("Restriction violated. Maximum retries exceeded.")
+            try:
+                llm_response = func(*args, **kwargs)
+                guard_result, violation_message = self.resolve_guard(llm_response)
+                if guard_result:
+                    self.handle_violation(violation_message)
+                else:
+                    return llm_response
+            except Exception as e:
+                self.retries = self.retries - 1
```

Comment:
@John-Church is the generator reset at each call ?

Reply:
sorry I'm not sure what you mean. I think yes? It's a new llm call each time